### PR TITLE
Header support

### PR DESCRIPTION
**Pull request by derekcollison**
This is first pass header support in the server. This introduces a new client protocol HPUB and bidirectional agreement between server and client on abilities using the header feature flag both in INFO and CONNECT.

Next pass will add in HMSG and complete propagation through the server with ability to strip headers for older clients.

/cc @nats-io/core

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -1404,6 +1407,9 @@ func (c *client) processConnect(arg []byte) error {
 	account := c.opts.Account
 	accountNew := c.opts.AccountNew
 	ujwt := c.opts.JWT
+	// For headers both need to support. If the server supports headers
+	/// it will have been set to true before this is called.
```
Comment (kozlovic) [inline server/client.go:1411]: extra "/"

**Comment by derekcollison** [comment]
I may put on basic HMSG and client stripping in small additional commit..

**Comment by derekcollison** [comment]
OK so I added proper support for client delivery and stripping for older clients. I did a bit of cleanup but fairly straightforward.

We do not currently care about contents of headers, but as I began thinking we will need to possibly at a future point for header injection and using headers to effect behavior.

Take a look when you can.

**Comment by derekcollison** [comment]
The next phase will be cross server. So no need to comment on lack of that support at this time.

File: server/server.go [hunk 2]
```diff
@@ -2254,6 +2253,17 @@ func (s *Server) ReadyForConnections(dur time.Duration) bool {
 	return false
 }
 
+// Quick utility to function to tell if the server supports headers.
+func (s *Server) supportsHeaders() bool {
+	if s == nil {
+		return false
+	}
+	s.mu.Lock()
```
Comment (kozlovic) [inline server/server.go:2261]: Either that (getting from s.info) or from options: `return !(s.getOpts().NoHeaderSupport)`

File: server/client.go [hunk 3]
```diff
@@ -3280,17 +3308,15 @@ func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, deliver,
...
-			msgh = append(msgh, sub.im.prefix...)
-			msgh = append(msgh, subj...)
-			msgh = append(msgh, ' ')
-			si = len(msgh)
+			dsubj = append(_dsubj[:0], sub.im.prefix...)
+			dsubj = append(dsubj, subj...)
 		}
 		// Normal delivery
-		mh := c.msgHeader(msgh[:si], sub, creply)
+		mh := c.msgHeader(dsubj, sub, creply)
```
[the remark this answers is not in the document]
Comment (derekcollison) [reply server/client.go:3319]: Yes for large fanouts we are doing more rewriting per client. I tried to keep the code flow but was a bit unwieldily. We did rewrites for some imports already but might revisit based on performance numbers.

File: server/server.go (hunk 2 above)
Comment (derekcollison) [reply server/server.go:2261]: That is better, will make that change.

Comment (derekcollison) [reply server/server.go]: Fixed, thanks.

**Comment by matthiashanel** [comment]
LGTM

**Comment by derekcollison** [comment]
Question for @matthiashanel and @kozlovic. Merge this and do server to server in new PR?

**Comment by kozlovic** [comment]
Separate I think.

**Pull request merged**